### ISSUE TITLE
nmsed_resp

### ISSUE THREAD

**Issue by ducha-aiki** [ducha-aiki/affnet#15]
What's the shape of `nmsed_resp` in the function [https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L260](NMS3dAndComposeA) ?  `nmsed_resp, idxs = torch.topk(nmsed_resp, k = num_features, dim = 0);`  the shape of input `nmsed_resp` should be `bachsize*3*height*width`. So, the shape of output `nmsed_resp` is `batchsize*3*height*k` ?  Is it right?
What's the role of the small piece of code? [https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L278](url)

Especially for the line `sc_y_x = F.conv2d(resp3d, self.grid,
                                padding = 1) / (F.conv2d(resp3d, self.grid_ones, padding = 1) + 1e-8)`

_Originally posted by @yunyundong in https://github.com/ducha-aiki/affnet/issues/13#issuecomment-427226751_

**Comment by ducha-aiki** [comment]
This keeps top k keypoints with strongest response. These are values of the responces of the keypoints, not their coordinates.  Output shape would be k x 1 

Regarding sc_y_x  - it contains scale, x and y of the keypoints. Kind of equivalent of the https://github.com/opencv/opencv_contrib/blob/master/modules/xfeatures2d/src/sift.cpp#L569-L572 same thing in SIFT detector.

Why it looks so different? Because in order to make it parallelizable and differentiable, everything should by expressed in the terms of conv2d or functions like this. So lots of classical code looks very strange. But otherwise there would be no point in implementing Hessian detector in PyTorch.

**Comment by yunyundong** [comment]
Strictly speaking, the `sc_y_x` calculated by `F.conv2d(resp3d, self.grid, padding = 1) / (F.conv2d(resp3d, self.grid_ones, padding = 1) + 1e-8)` is equal to the `xi, xr, and xc` in the opencv source code [Interpolate the location of Strong response points](https://github.com/opencv/opencv_contrib/blob/96c2ecb53af4a12247e0e2aee404b7f3a00a2aa7/modules/xfeatures2d/src/sift.cpp#L514-L518).? In opencv, its calculation fourmualr is  as the following picture shown: 
![image](https://user-images.githubusercontent.com/24907398/46569497-b1c4cc80-c988-11e8-9000-fb782b78e68e.png)

Besides, I am confused about the variable `img_scale` in the line [const float img_scale = 1.f/(255*SIFT_FIXPT_SCALE);](https://github.com/opencv/opencv_contrib/blob/96c2ecb53af4a12247e0e2aee404b7f3a00a2aa7/modules/xfeatures2d/src/sift.cpp#L480) from the source code of opencv,  its role is to normalize the value of `DoG` to range `[0,1]`? Thank you in advance.

**Comment by ducha-aiki** [comment]
Yes, is it similar, but not exactly. 
In OpenCV or this Hessian https://github.com/perdoch/hesaff/blob/master/pyramid.cpp#L158 exact subpixel point position is done by fitting 2nd order curve. It is better way, but harder to do in parallel.

I am using simpler "center of the mass" approach, which is implemented by two convilutions you mentioned. It is also used in LIFT
https://arxiv.org/abs/1603.09114
![image](https://user-images.githubusercontent.com/4803565/46570126-a3ae8600-c95f-11e8-8e63-62ef82345cdb.png)


Regarding img_scale, you are right.

**Comment by yunyundong** [comment]
Thank you very much, I got it. Thank you again.

**Comment by yunyundong** [comment]
Hi, @ducha-aiki , I have another problem, it is about the calculation of gradient of image in the `x` and `y` direction. Is some cases,  its kernel is `np.array([[[[0.5, 0, -0.5]]]]` as here [gx](https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L63), and  in other cases , its kernel is `np.array([[[[-1, 0, 1]]]]` [gx](https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L87) and [dx](https://github.com/perdoch/hesaff/blob/0d939a424372ed7dd23ecc0b3590f37122695cbc/affine.cpp#L24) from the hesaff, its kernel should be `np.array([[[[-1, 0, 1]]]]`.
Is there some standards to calculate the `dx`, or `dy`?  Especially for the second-derivates, such as [gxx](https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L69) and [gyy](https://github.com/ducha-aiki/affnet/blob/f46f0dcee547fb571c1c2d20adcd5e85cd6317b0/HandCraftedModules.py#L71-L72)

I do a simple example as following:

```
a=torch.randint(1,100,(1,10)).view(1,1,1,10)
gx =  nn.Conv2d(1, 1, kernel_size=(1,3), bias = False,padding=(0,1))
gx.weight.data = torch.from_numpy(np.array([[[[0.5, 0, -0.5]]]], dtype=np.float32))
gxx=nn.Conv2d(1, 1, kernel_size=(1,3), bias = False,padding=(0,1))
gxx.weight.data=torch.from_numpy(np.array([[[[1, -2, 1]]]], dtype=np.float32))
dxx1=gxx(a)
dxx2=gx(gx(a))
```
However, `dxx1!=dxx2`.   Especially for the cross-second derivate [dxy](https://github.com/perdoch/hesaff/blob/0d939a424372ed7dd23ecc0b3590f37122695cbc/pyramid.cpp#L141), its factor is `0.25`,  for [dxx, dyy](https://github.com/opencv/opencv_contrib/blob/96c2ecb53af4a12247e0e2aee404b7f3a00a2aa7/modules/xfeatures2d/src/sift.cpp#L500-L502), its factor is `1.0`.   Is there some derivation for the calculation about `dx, dy, dxx, dyy and dxy`?

**Comment by ducha-aiki** [comment]
@yunyundong 
Unfortunately, no standard way. There are two approaches: one is to incorporate normalization coefficient into the weights and the second - it to do it separately. 

You can read about this a bit here
https://cw.fel.cvut.cz/b172/courses/ucuss18/labs/00_intro

I also recommend next lab https://cw.fel.cvut.cz/b172/courses/ucuss18/labs/01_corresp

And corresponding slides-lectures
https://cw.fel.cvut.cz/b172/_media/courses/ucuss18/2018_local-features-orig_new_jc.pdf
